jujugui/static/gui/src/app/components/sharing/test-sharing.js

```diff
@@ -331,11 +334,83 @@ describe('Sharing', () => {
     const instance = renderer.getMountedInstance();
     instance._modifyModelAccessCallback('boom');
     const output = renderer.getRenderOutput();
-    const actualMessage = output.props.children[0].props.children[2];
+    const actualMessage = output.props.children[0].props.children[1].props.
+      children[3];
     const expectedMessage = (
-      <div className="sharing__invite--error">boom</div>
+      <div className="sharing__invite--error"><b>Error:</b>{' '}boom</div>
     );
     assert.deepEqual(actualMessage, expectedMessage);
   });
 
+  it('shows different buttons based on state', () => {
+    const renderer = jsTestUtils.shallowRender(
+      <juju.components.Sharing
+        addNotification={sinon.stub()}
+        canShareModel={true}
+        closeHandler={sinon.stub()}
+        getModelUserInfo={sinon.stub()}
+        grantModelAccess={sinon.stub()}
+        revokeModelAccess={sinon.stub()} />, true);
+    const instance = renderer.getMountedInstance();
+    let output = renderer.getRenderOutput();
+    const defaultAdd = output.props.children[0].
+      props.children[1].props.children[2].props.children;
+    const expectedDefaultAdd = (
+      <juju.components.GenericButton
+        submit={true}
+        title="Add"
+        tooltip="Add user"
+        ref="grantButton"
+        type="positive"
+        disabled={true}
+      />
+    );
+    assert.deepEqual(defaultAdd, expectedDefaultAdd);
+
+    instance.setState({canAdd: true});
```

Comment:
I'd rather see these pulled out into three separate tests; that way, if one button state breaks but the others work, we have 1 of 3 tests failing, rather than a failure within one monolithic test and we're not sure which state it was that broke.